press/press/doctype/deploy_candidate_build/deploy_candidate_build.py

```diff
@@ -0,0 +1,1169 @@
+		return find(
+			self.build_steps,
+			lambda x: x.stage_slug == stage_slug and x.step_slug == step_slug,
+		)
+
+	def _upload_build_context(self, context_filepath: str, build_server: str):
+		step = self.get_step("upload", "context") or frappe._dict()
+		step.status = "Running"
+		start_time = now()
+		self.save(ignore_version=True)
+
+		try:
+			upload_filename = self.upload_build_context_for_docker_build(
+				context_filepath,
+				build_server,
+			)
+		except Exception:
+			step.status = "Failure"
+			raise
+
+		step.status = "Success"
+		step.duration = get_duration(start_time)
+		self.save(ignore_version=True)
+		return upload_filename
+
+	def _package_build_context(self):
+		step = self.get_step("package", "context") or frappe._dict()
+		step.status = "Running"
+		start_time = now()
+		self.save(ignore_version=True)
+
+		def fix_content_permission(tarinfo):
+			tarinfo.uid = 1000
+			tarinfo.gid = 1000
+			return tarinfo
+
+		tmp_file_path = tempfile.mkstemp(suffix=".tar.gz")[1]
+		with tarfile.open(tmp_file_path, "w:gz", compresslevel=5) as tar:
+			if frappe.conf.developer_mode:
+				tar.add(self.build_directory, arcname=".", filter=fix_content_permission)
+			else:
+				tar.add(self.build_directory, arcname=".")
+
+		step.status = "Success"
+		step.duration = get_duration(start_time)
+		self.save(ignore_version=True)
+
+		return tmp_file_path
+
+	def _package_and_upload_context(self):
+		context_filepath = self._package_build_context()
+		context_filename = self._upload_build_context(
+			context_filepath,
+			self.build_server,
+		)
+		os.remove(context_filepath)
+		return context_filename
+
+	def _run_agent_jobs(self):
+		context_filename = self._package_and_upload_context()
+		settings = self.candidate._fetch_registry_settings()
+
+		Agent(self.build_server).run_build(
+			{
+				"filename": context_filename,
+				"image_repository": self.docker_image_repository,
+				"image_tag": self.docker_image_tag,
+				"registry": {
+					"url": settings.docker_registry_url,
+					"username": settings.docker_registry_username,
+					"password": settings.docker_registry_password,
+				},
+				"no_cache": self.no_cache,
+				"no_push": self.no_push,
+				# Next few values are not used by agent but are
+				# read in `process_run_build`
+				"deploy_candidate_build": self.name,
+				"deploy_after_build": self.deploy_after_build,
+			}
+		)
+
+		self.set_status(Status.RUNNING)
+
+	def _update_docker_image_metadata(self):
+		settings = self.candidate._fetch_registry_settings()
+
+		if settings.docker_registry_namespace:
+			namespace = f"{settings.docker_registry_namespace}/{settings.domain}"
+		else:
+			namespace = f"{settings.domain}"
+
+		self.docker_image_repository = f"{settings.docker_registry_url}/{namespace}/{self.group}"
+		self.docker_image_tag = self.name
+		self.docker_image = f"{self.docker_image_repository}:{self.docker_image_tag}"
+
+	def _start_build(self):
+		self._update_docker_image_metadata()
+		if self.no_build:
+			return
```

Comment:
This is redundant in `Deploy Candidate Build` doctype.
This thing should move to `Deploy Candidate`